src/utils/buildMetadata.ts

```diff
@@ -0,0 +1,254 @@
+            .map(({ id, name, searchable, mandatory, unique, valueType, displayInList, optionSet }) => ({
+                value: id,
+                text: name,
+                searchable,
+                mandatory,
+                unique,
+                valueType,
+                displayInList,
+                trackedEntityAttribute: { id },
+                optionSet,
+            }));
+
+        return { ...type, trackedEntityTypeAttributes };
+    });
+
+    const programDataElements = sheetProgramDataElements.map(dataElement => {
+        const optionSet = sheetOptionSets.find(({ name }) => name === dataElement.optionSet)?.id;
+
+        return {
+            ...dataElement,
+            domainType: "TRACKER",
+            optionSet: optionSet ? { id: optionSet } : undefined,
+        };
+    });
+
+    return {
+        dataSets,
+        dataElements: [...dataElements, ...programDataElements],
+        options,
+        sections,
+        categories,
+        categoryCombos,
+        categoryOptions,
+        optionSets,
+        trackedEntityAttributes,
+        trackedEntityTypes,
+        //programs,
+        programRules: buildProgramRules(sheets),
+        programRuleActions: buildProgramRuleActions(sheets),
+        programRuleVariables: buildProgramRuleVariables(sheets),
+    };
+}
+
+function buildProgramRules(sheets: Sheet[]) {
+    const rules = getItems(sheets, "programRules");
+    const programs = getItems(sheets, "programs");
+    const actions = getItems(sheets, "programRuleActions");
+
+    return rules.map(rule => {
+        const program = getByName(programs, rule.program);
+
+        const programRuleActions = actions
+            .filter(action => action.programRule === rule.name)
+            .map(action => ({ id: action.id }));
+
+        return { ...rule, program: { id: program.id }, programRuleActions };
+    });
+}
+
+function buildProgramRuleActions(sheets: Sheet[]) {
+    const get = (name: string) => getItems(sheets, name); // shortcut
+
+    const actions = get("programRuleActions"),
+        rules = get("programRules"),
+        elements = get("programDataElements"),
+        attrs = get("trackedEntityAttributes"),
+        stages = get("programStages"),
+        sections = get("programStageSections");
+
+    return actions.map(action => {
+        let data = { ...action } as MetadataItem; // copy that we will modify
+
+        replaceById(data, "programRule", rules);
+        replaceById(data, "dataElement", elements);
+        replaceById(data, "trackedEntityAttribute", attrs);
+        replaceById(data, "programStage", stages);
+        replaceById(data, "programStageSection", sections);
+
+        return data;
+    });
+}
+
+function buildProgramRuleVariables(sheets: Sheet[]) {
+    const get = (name: string) => getItems(sheets, name); // shortcut
+
+    const vars = get("programRuleVariables"),
+        programs = get("programs"),
+        elements = get("programDataElements"),
```

Comment:
This should point to programStageDataElements, not programDataElements.

Reply:
So, after talking with Jordi, there was a misunderstanding on my part about programDataElements.
That spreadsheet tab is not for the metadata type programDataElements, instead it's for dataElements of  TRACKER domainType.
I'm not sure if metadata type programDataElements is needed for metadata work (or what it is), the documentation is quite sparse about it.
Anyway, with this issue resolved there is no need for these changes.